File: backend/src/main/java/entities/Product.java
package entities;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Size;
import lombok.*;
import java.util.*;

import static jakarta.persistence.FetchType.EAGER;
import static jakarta.persistence.FetchType.LAZY;

@Data
@Entity
@Table(name = "product")
public class Product {
    @Id
    @NotNull
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @Basic
    @Size(max = 15)
    @NotNull
    @Column(name = "bar code")
    private String barCode;

    @Basic
    @Size(max = 30)
    @NotNull
    @Column(name = "name")
    private String name;

    @NotNull
    @ManyToOne(fetch = EAGER)
    @JoinColumn(name = "category")
    private Category category;

    @Basic
    @NotNull
    @Column(name = "price")
    private Double price;

    @Basic
    @Size(max = 200)
    @Column(name = "description")
    @ToString.Exclude
    @JsonIgnore
    private String description;

    @Basic
    @NotNull
    @Column(name = "quantity")
    private Integer quantity;

    @Version
    @NotNull
    @Column(name = "version")
    @JsonIgnore
    private long version;

    @JsonIgnore
    @ToString.Exclude
    @OneToMany(fetch = LAZY,mappedBy = "product",cascade = CascadeType.ALL, orphanRemoval = true)
    private Set<OrderProduct> orderProducts = new LinkedHashSet<>();

}

File: frontend/src/app/models/Product.ts
export interface Product {
  id: number;
  barCode: string;
  name: string;
  category: string;
  price: number;
  description: string;
  quantity: number;
}
export class ProductImpl implements Product{
  public id: number;
  public barCode: string;
  public name: string;
  public category: string;
  public price: number;
  public description: string;
  public quantity: number;

  constructor(){
    this.id=1;
    this.barCode='';
    this.name='';
    this.category='';
    this.price=0.0;
    this.description='';
    this.quantity=0;
  }

  copy(p:Product) : ProductImpl{
    this.id=p.id;
    this.barCode=p.barCode;
    this.name=p.name;
    this.price=p.price;
    this.description=p.description;
    this.quantity=p.quantity;
    return this;
  }


}
